src/server/json_family.cc

```diff
@@ -164,8 +202,159 @@ OpResult<vector<OptSizeT>> OpStrLen(const OpArgs& op_args, string_view key, Json
+    return it_res.status();
+
+  error_code ec;
+  json_decoder<json> decoder;
+  const PrimeValue& pv = it_res.value()->second;
+
+  string val = GetString(op_args.shard, pv);
+  basic_json_parser<char> parser(basic_json_decode_options<char>{}, &JsonErrorHandler);
+
+  parser.update(val);
+  parser.finish_parse(decoder, ec);
+
+  if (!decoder.is_valid()) {
+    return OpStatus::SYNTAX_ERR;
+  }
+
+  vector<OptSizeT> vec;
+  auto cb = [&vec](const string_view& path, const json& val) {
+    if (val.is_array()) {
+      vec.emplace_back(val.array_value().size());
+    } else {
+      vec.emplace_back(nullopt);
+    }
+  };
+
+  expression.evaluate(decoder.get_result(), cb);
+  return vec;
+}
+
+OpResult<vector<OptBool>> OpToggle(const OpArgs& op_args, string_view key, string_view path) {
+  auto& db_slice = op_args.shard->db_slice();
+  OpResult<PrimeIterator> it_res = db_slice.Find(op_args.db_ind, key, OBJ_STRING);
+  if (!it_res.ok())
+    return it_res.status();
+
+  error_code ec;
+  json_decoder<json> decoder;
+  const PrimeValue& pv = it_res.value()->second;
+
+  string val = GetString(op_args.shard, pv);
+  basic_json_parser<char> parser(basic_json_decode_options<char>{}, &JsonErrorHandler);
+
+  parser.update(val);
+  parser.finish_parse(decoder, ec);
+
+  if (!decoder.is_valid()) {
+    return OpStatus::SYNTAX_ERR;
+  }
+
+  vector<OptBool> vec;
+  auto cb = [&vec](const string& path, json& val) {
+    if (val.is_bool()) {
+      bool current_val = val.as_bool() ^ true;
+      val = current_val;
+      vec.emplace_back(current_val);
+    } else {
+      vec.emplace_back(nullopt);
+    }
+  };
+
+  json j = decoder.get_result();
+  ec = json_replace_with_ec(j, path, cb);
+  if (ec) {
+    VLOG(1) << "Failed to evaulate expression on json with error: " << ec.message();
+    return OpStatus::SYNTAX_ERR;
+  }
+
+  auto [it_output, added] = db_slice.AddOrFind(op_args.db_ind, key);
+  it_output->second.SetString(j.as_string());
+  db_slice.PostUpdate(op_args.db_ind, it_output);
+  RecordJournal(op_args, it_output->first, it_output->second);
+
+  return vec;
+}
+
 } // namespace
 
+void JsonFamily::Toggle(CmdArgList args, ConnectionContext* cntx) {
+  string_view key = ArgS(args, 1);
+  string_view path = ArgS(args, 2);
+
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    return OpToggle(t->GetOpArgs(shard), key, path);
+  };
+
+  DVLOG(1) << "Before Get::ScheduleSingleHopT " << key;
+  Transaction* trans = cntx->transaction;
+  OpResult<vector<OptBool>> result = trans->ScheduleSingleHopT(move(cb));
+
+  if (result) {
+    DVLOG(1) << "JSON.TOGGLE " << trans->DebugId() << ": " << key;
+    if (result->empty()) {
+      (*cntx)->SendNullArray();
+    } else {
+      (*cntx)->StartArray(result->size());
+      for (auto& it: *result) {
+        if (it.has_value()) {
+          (*cntx)->SendLong(*it);
+        } else {
+          (*cntx)->SendNull();
+        }
+      }
+    }
+  } else {
+    switch (result.status()) {
```

Comment:
I think if you just call `(*cntx)->SendError(result.status())` you will get a behavior equivalent to that switch.   
if not, please tell me where it breaks.